EVM Bridge Docs (#43)
ripple/opensource.ripple.com: docs/evm-sidechain/deploy-a-bridge.md

---
html: deploy-a-bridge.html
blurb: Follow these steps to deploy a bridge.
labels:
  - Interoperability
status: not_enabled
---
# Deploy a Bridge

Before deploying a bridge, familiarize yourself with the parameters and terms used:

- **Currency:** The currency used by the bridge. Depending on the token, the bridge has different setups. If the currency is XRP, the door account is called a _Native Bridge Door_; if the currency is a token, the door account is called a _Token Bridge Door_.

> - **Locking Chain:** The chain where the token originally exists. If the locking chain is an XRPL chain, the bridge IOU issuer isn't the bridge door account. If the locking chain is an EVM chain, the ERC20 owner isn't the bridge door account.

Reviewer: We try to avoid the term "IOU". Also, I'd rather tell the user what the issuer _is_ rather than what it isn't.

```suggestion
- **Locking Chain:** The chain where the token originally exists. If the locking chain is an XRPL chain, the token has an issuer that is separate from the bridge door account. If the locking chain is an EVM chain, the token has an ERC20 owner that is separate from the bridge door account.
```

> - **Issuing Chain:** The chain the token is bridged to. If the issuing chain is an XRPL chain, the bridged IOU issuer is the bridge door account. If the issuing chain is an EVM chain, the bridged ERC20 owner is the bridge door account and can mint/burn tokens.

Reviewer: Prefer "token" to IOU here too.

- **Owners:** The witnesses addresses. For each witness, you must have its address on both chains.
- **Threshold:** The amount of attestations needed for each operation.

> - **Signautre Reward:** The reward distributed to witnesses for providing attestations.

Reviewer: ```suggestion
- **Signature Reward:** The reward distributed to witnesses for providing attestations.
```

> - **Minimum Create Account:** For native XRP bridges, this parameter sets the minimum amount of XRP that needs to be locked in order to execute a create account operation.

Reviewer: ```suggestion
- **Minimum Create Account:** For native XRP bridges, this parameter sets the minimum amount of XRP needed to create an account on the issuing chain.
```

## Create a Safe Account

Before deploying a bridge, you have to create a Gnosis Safe. The safe is the multisig account that manages all of the bridge's assets.

> <iframe width="560" height="315" src="https://www.youtube.com/embed/s9J0516Ta64" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

Reviewer: Maybe add a lead-in or caption for the Youtube video?

1. Connect a wallet to the [Gnosis Safe management platform](https://safe-evm-poa-sidechain.peersyst.tech/).

> ![Connect a Wallet](../../images/evm-connect-wallet.png "Connect a Wallet")

Reviewer: Better alt text would make this more useful. For example,

```suggestion
![Screenshot of the "Connect a Wallet" interface in the Gnosis Safe management platform. A "Connect a Wallet" navigation item with a "Connect" button is visible underneath an overlay that asks the user to select from "Available Wallets". The available wallets include MetaMask, Safe Mobile, WalletConnect, and others.](../../images/evm-connect-wallet.png "Connect a Wallet")
```

But also, this seems to actually be several steps:

1. Connect to the EVM Sidechain's instance of Gnosis Safe (at the given link)
2. Click the tiny "Connect a Wallet" text in the top navbar
3. Click the "Connect" button in the popup that appears
4. Select a wallet type from the available options
5. Complete the specific steps for the selected wallet type

2. Click **+ Create New Safe**.

![Create New Safe](../../images/evm-create-safe.png "Create New Safe")

- Give the safe a name to identify it locally.
- The owners are the bridge witnesses.
- The threshold is the amount of signatures required to make a transaction on the bridge.

Before submitting, check your information on the review screen.

> ![Review Safe Info](../../images/evm-safe-owners.png "Review Safe Info")

Reviewer: Some useful information that might be helpful to pull out of this screenshot include the example values, that the owners have hex values (public keys?) associated with them, and the example threshold is 2 out of 3.

> A prompt will appear, asking to sign the transaction. Carefully review the transaction details before signing it. Once the transaction is submitted, you'll see a success screen.

Reviewer: I could be wrong here, but I think the prompt comes from your wallet app, so:

```suggestion
Your wallet displays a prompt asking to sign the transaction. Carefully review the transaction details before signing it. After submitting the transaction, a success screen displays.
```

![Safe Created](../../images/evm-safe-created.png "Safe Created")

## Deploy the Bridge Doors

Depending on the bridge currency you want to deploy, there are different steps. Choose the correct bridge type below:

### XRP Native Bridge

> XRP bridges are a rare case and should only have one bridge per mainchain and sidechain pair. The sidechain specification forces the XRP issuer account to be `rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh`.

Reviewer: I think it's unnecessarily speculative to call it a rare case. Also, I think you can reduce the amount of sidechain-related jargon in this sentence:

```suggestion
There should should only be at most one XRP native bridge between any two chains. The sidechain specification requires the XRP issuer account to be `rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh`.
```

> To deploy an XRP bridge door smart contract, run this typescript command:

Reviewer: This isn't a command, it's a whole script.

```suggestion
To deploy an XRP bridge door smart contract, run a script such as the following example, filling in the constants with the appropriate values for your situation:
```

```typescript
import { BridgeDoorNative__factory } from "@peersyst/xrp-evm-contracts/dist/typechain/factories/BridgeDoorNative__factory";
import { xrplAccountToEvmAddress } from "./util/address-derivation";
import { ethers } from "ethers";

const safeAddress = "0x...";
const lockingBridgeAddress = "r1..."; // Address of the XRPL bridge
const minRewardAmount = "1";
const minCreateAmount = "50";

const signerWallet = new ethers.Wallet("0..."); // Private key with balance

const bridgeDoorNativeFactory = new BridgeDoorNative__factory(signerWallet);
const bridgeDoorNative = await bridgeDoorNativeFactory.deploy(
    safeAddress,
    ethers.utils.parseEther(minRewardAmount),
    ethers.utils.parseEther(minCreateAmount),
    ethers.constants.AddressZero,
    xrplAccountToEvmAddress(lockingBridgeAddress),
);
const bridgeAddress = bridgeDoorNative.address;
```

### Token Bridge

> Token bridges are more common than XRP bridges, so you'll most likely deploy this type.

Reviewer: Consider removing this entirely.

> To deploy a token bridge door smart contract, run this typescript command on the locking chain:

Reviewer: Again, this isn't a "command".
